File: 单调栈/5. 柱状图中最大柱形 - 84/Largest Rectangle in Histogram.cpp
/*
    23/03/27
    https://leetcode.cn/problems/largest-rectangle-in-histogram/
 */

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

/*
    solution: 栈
    result: ✅
    TC: O(n)
    SC: O(n)
 */

int largestRectangleArea(vector<int>& heights) {
    
    heights.insert(heights.begin(), 0);
    heights.push_back(0);
    stack<int> st;
    st.push(0);
    int result = 0;
    int lengthOfHeights = heights.size();

    for (int i = 1; i < lengthOfHeights; i++) {
        while (heights[i] < heights[st.top()]) {
            int mid_height = st.top();
            st.pop();
            int w = i - st.top() - 1;
            int h = heights[mid_height];
            result = max(result, w*h);
        }
        st.push(i);
    }

    return result;
}

File: 单调栈/5. 柱状图中最大柱形 - 84/Largest Rectangle in Histogram.py
# 23/03/27
# https://leetcode.cn/problems/largest-rectangle-in-histogram/submissions/

from collections import List

# 栈
# ✅

def largestRectangleArea(self, heights: List[int]) -> int:

    heights = [0] + heights
    heights = heights +  [0]
    st = [0]
    result = 0
    lengthOfHeight = len(heights)

    for i in range(1, lengthOfHeight):
        while heights[i] < heights[st[-1]]:
            mid = st.pop()
            w = i - st[-1] - 1
            h = heights[mid]
            result = max(result, w * h)
        st.append(i)
    
    return result